### PR TITLE
NIO Support

### PR DESCRIPTION
**Pull request by stephencelis**
This is a proof-of-concept and has some stuff to figure out.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -14,12 +14,16 @@ xcodeproj:
...
 	xed .
 
-bootstrap: common-crypto-mm xcodeproj
+bootstrap: common-crypto-mm xcodeproj xcodeproj-mm
 
 common-crypto-mm:
 	-@sudo mkdir -p "$(COMMON_CRYPTO_PATH)"
 	-@echo "$$COMMON_CRYPTO_MODULE_MAP" | sudo tee "$(COMMON_CRYPTO_MODULE_MAP_PATH)" > /dev/null
 
+xcodeproj-mm:
```
Comment (stephencelis) [inline Makefile]: This is a fun one. It installs any module maps that are generated with the Xcode project.

File: Sources/HttpPipeline/NIO.swift [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    let reqPart = self.unwrapInboundIn(data)
+
+    switch reqPart {
+    case let .head(header):
+      self.request = URL(string: header.uri).map { url -> URLRequest in
+        var req = URLRequest(url: url)
+          |> \.httpMethod .~ method(from: header.method)
+          |> \.allHTTPHeaderFields .~ header.headers.reduce(into: [:]) { $0[$1.name] = $1.value }
+        req.url = req.value(forHTTPHeaderField: "Host")
+          .flatMap { URL(string: "http://" + $0 + url.absoluteString) }
```
Comment (stephencelis) [inline Sources/HttpPipeline/NIO.swift]: This logic is almost certainly wrong. Need to figure out how to properly construct the full request URL.

File: Sources/HttpPipeline/NIO.swift [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+      _ = ctx.channel.write(HTTPServerResponsePart.body(.byteBuffer(buffer)))
+
+      _ = ctx.channel.writeAndFlush(HTTPServerResponsePart.end(nil)).then {
+        ctx.channel.close()
+      }
+    }
+  }
+}
+
+private func method(from method: HTTPMethod) -> String {
```
Comment (stephencelis) [inline Sources/HttpPipeline/NIO.swift]: Maybe I'm missing something, but there didn't seem to be a way to get the string representation of this thing.

File: Sources/HttpPipeline/NIO.swift [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+  do {
+    let reuseAddrOpt = ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR)
+    let group = MultiThreadedEventLoopGroup(numThreads: System.coreCount)
+    let bootstrap = ServerBootstrap(group: group)
+      .serverChannelOption(ChannelOptions.backlog, value: 256)
+      .serverChannelOption(reuseAddrOpt, value: 1)
+      .childChannelInitializer { channel in
+        channel.pipeline.configureHTTPServerPipeline()
+          .then { channel.pipeline.add(handler: HTTPResponseCompressor()) }
```
Comment (stephencelis) [inline Sources/HttpPipeline/NIO.swift]: This enables gzipped responses. Not sure if it's worth making this configurable or not or if it should be disabled in dev. Seems fine?

File: Makefile (hunk 1 above)
Comment (mbrandonw) [reply Makefile]: oh nice, can that replace our other modulemap copying?

File: Package.swift [hunk 5]
```diff
@@ -16,14 +16,16 @@ let package = Package(
     .library(name: "HtmlTestSupport", targets: ["HtmlTestSupport"]),
     .library(name: "HtmlPrettyPrint", targets: ["HtmlPrettyPrint"]),
     .library(name: "HttpPipeline", targets: ["HttpPipeline"]),
+//    .executable(name: "HttpPipelineExample", targets: ["HttpPipelineExample"]),
```
Comment (mbrandonw) [inline Package.swift]: do we wanna leave these in?

Comment (stephencelis) [reply Package.swift]: I'd like to get it workin but didn't want it to block this PR.

File: Sources/HttpPipeline/NIO.swift [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  case .PROPPATCH: return "PROPPATCH"
+  case .SUBSCRIBE: return "SUBSCRIBE"
+  case .MKCALENDAR: return "MKCALENDAR"
+  case .MKACTIVITY: return "MKACTIVITY"
+  case .UNSUBSCRIBE: return "UNSUBSCRIBE"
+  case let .RAW(value): return value
+  }
+}
+
+private func headers(from headers: HTTPHeaders) -> [String: String] {
```
Comment (mbrandonw) [inline Sources/HttpPipeline/NIO.swift]: this doesn't look used anymore

Comment (stephencelis) [reply Sources/HttpPipeline/NIO.swift]: Whoops, yeah. Deleted on the branch I had goin before 4.1 but forgot to update this one.

**Pull request merged**